src/snowflake/snowpark/functions.py

```diff
@@ -8700,3 +8704,139 @@ def make_interval(
+        >>> new_df.show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        ---------
+        <BLANKLINE>
+
+    Example 2::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()])
+        >>> new_df.show()
+        -----------------
+        |"C_1"  |"C_2"  |
+        -----------------
+        |a      |30     |
+        -----------------
+        <BLANKLINE>
+
+    Example 3::
+
+        >>> new_df = map(df, lambda row: (row[1], row[0] * 3), output_types=[StringType(), IntegerType()], output_column_names=['col1', 'col2'])
+        >>> new_df.show()
+        -------------------
+        |"COL1"  |"COL2"  |
+        -------------------
+        |a       |30      |
+        -------------------
+        <BLANKLINE>
+
+    Example 4::
+        >>> df = session.create_dataframe([10, 20, 30])
+        >>> new_df = map(df, lambda x: x * x, output_types=[IntegerType()], wrap_row=False)
+        >>> new_df.sort(new_df[0]).show()
+        ---------
+        |"C_1"  |
+        ---------
+        |100    |
+        |400    |
+        |900    |
+        ---------
+        <BLANKLINE>
+
+    Note:
+        The result of the `func` function must be either a scalar value or a tuple containing the same number of elements
+        as specified in the `output_types` argument.
+    """
+
+    kwargs["packages"] = add_snowpark_package_to_sproc_packages(
+        dataframe._session, kwargs.get("packages")
+    )
+    if len(output_types) == 0:
+        raise ValueError("output_types cannot be empty.")
+    input_types = [field.datatype for field in dataframe.schema.fields]
+    num_fields = len(dataframe.schema.fields)
+
+    if output_column_names is None:
+        output_column_names = [f"c_{i+1}" for i in range(len(output_types))]
+    elif len(output_column_names) != len(output_types):
+        raise ValueError(
+            "'output_column_names' and 'output_types' must be of the same size."
+        )
+
+    output_schema = [
+        StructField(name, type_)
+        for name, type_ in zip(output_column_names, output_types)
+    ]
+    output_cols = [
+        col(f"${i + 1}") for i in range(num_fields, num_fields + len(output_types))
+    ]
+    input_row_obj = snowflake.snowpark.Row(*dataframe.columns) if wrap_row else None
+
+    def wrap_result(result):
+        if isinstance(result, list) or isinstance(result, snowflake.snowpark.Row):
+            return tuple(result)
+        elif isinstance(result, tuple):
+            return result
+        else:
+            return (result,)
+
+    class _MapFunc:
+        def process(self, *argv):
+            if wrap_row:
+                argv = [input_row_obj(*argv)]
+            result = func(*argv)
+            yield wrap_result(result)
+
+    map_udtf = dataframe._session.udtf.register(
+        _MapFunc,
+        output_schema=StructType(output_schema),
+        input_types=input_types,
+        is_permanent=False,
```

Comment:
let user decide if they want this registration to be permanent. If would allow users to not do repeated registration since it is a time consuming process.

Reply:
Applied in a new commit.